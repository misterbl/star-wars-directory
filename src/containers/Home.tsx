import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { IHomeComponent } from "./Home.d";
import ResultList from "../components/ResultList";
import SelectFilmForm, { ISelectFilmForm } from "../components/SelectFilmForm";
import SelectGenderForm, {
  ISelectGenderForm
} from "../components/SelectGenderForm";
import { getPeopleList, getFilmsList } from "../selectors/appSelectors";
import { IAppState } from "../state";
import { Formik } from "formik";

export class Home extends Component<IHomeComponent> {
  state = {
    showPeople: true,
    showFilms: true,
    peopleList: []
  };

  componentDidMount() {
    this.setState({ peopleList: this.props.peopleList });
  }
  pushToResistance = () => {
    window.location.href =
      "https://www.starwars.com/tv-shows/star-wars-resistance";
  };

  changePeopleView = () => {
    this.setState({ showPeople: !this.state.showPeople });
  };

  changeFilmsView = () => {
    this.setState({ showFilms: !this.state.showFilms });
  };

  handleFilmChange = async (event: any) => {
    console.log(event.selectedFilm);
    const { peopleList } = this.props;
    console.log(peopleList);

    await this.setState({
      peopleList: peopleList.filter(character => {
        character.films.indexOf(event.selectedFilm) !== -1;
      })
    });

    const result =
      (await peopleList.length) > 0 &&
      peopleList.filter(character => {
        character.films.indexOf(event.selectedFilm) !== -1;
      });
    console.log(result);
    console.log(this.state);
  };

  handleGenderChange = async (event: any) => {
    // console.log(event.selectedGender);
    // const { peopleList } = this.state;
    // await this.setState({
    //   peopleList: peopleList.filter(
    //     character => character.gender === event.selectedGender
    //   )
    // });
    console.log(this.state);
  };

  render() {
    const { showPeople, showFilms } = this.state;
    const { filmsList, peopleList } = this.props;
    const foundPeople = peopleList.length > 0;
    const foundFilms = filmsList.length > 0;
    const resultLength = peopleList.length + filmsList.length;
    return (
      <main className="form-container p-5">
        <div className="resistance mb-4" onClick={this.pushToResistance} />
        {foundPeople && foundFilms && (
          <div className="result-filter my-4 p-3">
            <div className="text-white">{`${resultLength} RESULTS`}</div>
            <label htmlFor="showPeople">
              Show characters
              <input
                type="checkbox"
                id="showPeople"
                checked={showPeople}
                onChange={this.changePeopleView}
              />
            </label>
            <br />
            <label htmlFor="showFilms">
              Show films
              <input
                type="checkbox"
                id="showFilms"
                checked={showFilms}
                onChange={this.changeFilmsView}
              />
            </label>
          </div>
        )}
        {foundPeople && showPeople && (
          <>
            <Formik<ISelectFilmForm>
              initialValues={{
                selectedFilm: ""
              }}
              onSubmit={this.handleFilmChange}
              render={formikProps => <SelectFilmForm {...formikProps} />}
            />
            <Formik<ISelectGenderForm>
              initialValues={{
                selectedGender: ""
              }}
              onSubmit={this.handleGenderChange}
              render={formikProps => <SelectGenderForm {...formikProps} />}
            />
            <ResultList type="Characters" list={this.props.peopleList} />
          </>
        )}
        {foundFilms && showFilms && (
          <ResultList type="Films" list={filmsList} />
        )}
      </main>
    );
  }
}

export const mapStateToProps = (state: IAppState) => ({
  peopleList: getPeopleList(state),
  filmsList: getFilmsList(state)
});

export default withRouter(connect(mapStateToProps)(Home));
