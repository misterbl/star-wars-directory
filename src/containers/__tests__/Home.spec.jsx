/* eslint-disable no-undef */
import * as React from "react";
import { shallow } from "enzyme";
import history from "../../testMocks/history.mock";
import { Home } from "../Home";
import ROUTES from "../../routes";

describe("Home", () => {
  const props = {
    history,
    peopleList: [
      {
        name: "name",
        height: "123",
        mass: "12",
        birth_year: "1234",
        gender: "male",
        hair_color: "hair_color",
        eye_color: "eye_color"
      },
      {
        name: "name2",
        height: "12333",
        mass: "12333",
        birth_year: "1234232",
        gender: "female",
        hair_color: "black",
        eye_color: "black"
      }
    ],
    filmsList: ["films1", "films2"]
  };
  const wrapper = shallow(<Home {...props} />);
  it("matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("update the state on ComponentDidMount", () => {
    expect(wrapper.state().peopleList).toEqual(props.peopleList);
  });
  it("update the state on checkbox showPeople check", () => {
    const showPeople = wrapper.find("#showPeople");
    showPeople.simulate("change");
    expect(wrapper.state().showPeople).toBeFalsy();
  });
  it("update the state on checkbox showFilms check", () => {
    const showFilms = wrapper.find("#showFilms");
    showFilms.simulate("change");
    expect(wrapper.state().showFilms).toBeFalsy();
  });
  it("doesn't show the checkbox showFilms nor showPeople if peopleList is empty ", () => {
    const newProps = { ...props, peopleList: [] };
    const newWrapper = shallow(<Home {...newProps} />);
    const showFilms = newWrapper.find("#showFilms");
    const showPeople = newWrapper.find("#showPeople");
    expect(showFilms.length).toEqual(0);
    expect(showPeople.length).toEqual(0);
  });
});
