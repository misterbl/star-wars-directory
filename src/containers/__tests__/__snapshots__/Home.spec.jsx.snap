// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    filmsList={
      Array [
        "films1",
        "films2",
      ]
    }
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 123,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "",
          "state": "",
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    peopleList={
      Array [
        Object {
          "birth_year": "1234",
          "eye_color": "eye_color",
          "gender": "male",
          "hair_color": "hair_color",
          "height": "123",
          "mass": "12",
          "name": "name",
        },
        Object {
          "birth_year": "1234232",
          "eye_color": "black",
          "gender": "female",
          "hair_color": "black",
          "height": "12333",
          "mass": "12333",
          "name": "name2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="resistance mb-4"
          onClick={[Function]}
        />,
        <React.Fragment>
          <label
            className="text-white"
            htmlFor="showPeople"
          >
            Show characters
          </label>
          <input
            checked={true}
            id="showPeople"
            onClick={[Function]}
            type="checkbox"
          />
          <br />
          <label
            className="text-white"
            htmlFor="showFilms"
          >
            Show films
          </label>
          <input
            checked={true}
            id="showFilms"
            onClick={[Function]}
            type="checkbox"
          />
          <div
            className="text-white my-3"
          >
            4 RESULTS
          </div>
        </React.Fragment>,
        <React.Fragment>
          <Formik
            enableReinitialize={false}
            initialValues={
              Object {
                "selectedFilm": "",
              }
            }
            isInitialValid={false}
            onSubmit={[Function]}
            render={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
          />
          <Formik
            enableReinitialize={false}
            initialValues={
              Object {
                "selectedGender": "",
              }
            }
            isInitialValid={false}
            onSubmit={[Function]}
            render={[Function]}
            validateOnBlur={true}
            validateOnChange={true}
          />
          <ResultList
            list={
              Array [
                Object {
                  "birth_year": "1234",
                  "eye_color": "eye_color",
                  "gender": "male",
                  "hair_color": "hair_color",
                  "height": "123",
                  "mass": "12",
                  "name": "name",
                },
                Object {
                  "birth_year": "1234232",
                  "eye_color": "black",
                  "gender": "female",
                  "hair_color": "black",
                  "height": "12333",
                  "mass": "12333",
                  "name": "name2",
                },
              ]
            }
            type="Characters"
          />
        </React.Fragment>,
        <ResultList
          list={
            Array [
              "films1",
              "films2",
            ]
          }
          type="Films"
        />,
      ],
      "className": "form-container p-5",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "resistance mb-4",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <label
              className="text-white"
              htmlFor="showPeople"
            >
              Show characters
            </label>,
            <input
              checked={true}
              id="showPeople"
              onClick={[Function]}
              type="checkbox"
            />,
            <br />,
            <label
              className="text-white"
              htmlFor="showFilms"
            >
              Show films
            </label>,
            <input
              checked={true}
              id="showFilms"
              onClick={[Function]}
              type="checkbox"
            />,
            <div
              className="text-white my-3"
            >
              4 RESULTS
            </div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Show characters",
              "className": "text-white",
              "htmlFor": "showPeople",
            },
            "ref": null,
            "rendered": "Show characters",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": true,
              "id": "showPeople",
              "onClick": [Function],
              "type": "checkbox",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Show films",
              "className": "text-white",
              "htmlFor": "showFilms",
            },
            "ref": null,
            "rendered": "Show films",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": true,
              "id": "showFilms",
              "onClick": [Function],
              "type": "checkbox",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "4 RESULTS",
              "className": "text-white my-3",
            },
            "ref": null,
            "rendered": "4 RESULTS",
            "type": "div",
          },
        ],
        "type": Symbol(react.fragment),
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Formik
              enableReinitialize={false}
              initialValues={
                Object {
                  "selectedFilm": "",
                }
              }
              isInitialValid={false}
              onSubmit={[Function]}
              render={[Function]}
              validateOnBlur={true}
              validateOnChange={true}
            />,
            <Formik
              enableReinitialize={false}
              initialValues={
                Object {
                  "selectedGender": "",
                }
              }
              isInitialValid={false}
              onSubmit={[Function]}
              render={[Function]}
              validateOnBlur={true}
              validateOnChange={true}
            />,
            <ResultList
              list={
                Array [
                  Object {
                    "birth_year": "1234",
                    "eye_color": "eye_color",
                    "gender": "male",
                    "hair_color": "hair_color",
                    "height": "123",
                    "mass": "12",
                    "name": "name",
                  },
                  Object {
                    "birth_year": "1234232",
                    "eye_color": "black",
                    "gender": "female",
                    "hair_color": "black",
                    "height": "12333",
                    "mass": "12333",
                    "name": "name2",
                  },
                ]
              }
              type="Characters"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "enableReinitialize": false,
              "initialValues": Object {
                "selectedFilm": "",
              },
              "isInitialValid": false,
              "onSubmit": [Function],
              "render": [Function],
              "validateOnBlur": true,
              "validateOnChange": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "enableReinitialize": false,
              "initialValues": Object {
                "selectedGender": "",
              },
              "isInitialValid": false,
              "onSubmit": [Function],
              "render": [Function],
              "validateOnBlur": true,
              "validateOnChange": true,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "list": Array [
                Object {
                  "birth_year": "1234",
                  "eye_color": "eye_color",
                  "gender": "male",
                  "hair_color": "hair_color",
                  "height": "123",
                  "mass": "12",
                  "name": "name",
                },
                Object {
                  "birth_year": "1234232",
                  "eye_color": "black",
                  "gender": "female",
                  "hair_color": "black",
                  "height": "12333",
                  "mass": "12333",
                  "name": "name2",
                },
              ],
              "type": "Characters",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Symbol(react.fragment),
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "list": Array [
            "films1",
            "films2",
          ],
          "type": "Films",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="resistance mb-4"
            onClick={[Function]}
          />,
          <React.Fragment>
            <label
              className="text-white"
              htmlFor="showPeople"
            >
              Show characters
            </label>
            <input
              checked={true}
              id="showPeople"
              onClick={[Function]}
              type="checkbox"
            />
            <br />
            <label
              className="text-white"
              htmlFor="showFilms"
            >
              Show films
            </label>
            <input
              checked={true}
              id="showFilms"
              onClick={[Function]}
              type="checkbox"
            />
            <div
              className="text-white my-3"
            >
              4 RESULTS
            </div>
          </React.Fragment>,
          <React.Fragment>
            <Formik
              enableReinitialize={false}
              initialValues={
                Object {
                  "selectedFilm": "",
                }
              }
              isInitialValid={false}
              onSubmit={[Function]}
              render={[Function]}
              validateOnBlur={true}
              validateOnChange={true}
            />
            <Formik
              enableReinitialize={false}
              initialValues={
                Object {
                  "selectedGender": "",
                }
              }
              isInitialValid={false}
              onSubmit={[Function]}
              render={[Function]}
              validateOnBlur={true}
              validateOnChange={true}
            />
            <ResultList
              list={
                Array [
                  Object {
                    "birth_year": "1234",
                    "eye_color": "eye_color",
                    "gender": "male",
                    "hair_color": "hair_color",
                    "height": "123",
                    "mass": "12",
                    "name": "name",
                  },
                  Object {
                    "birth_year": "1234232",
                    "eye_color": "black",
                    "gender": "female",
                    "hair_color": "black",
                    "height": "12333",
                    "mass": "12333",
                    "name": "name2",
                  },
                ]
              }
              type="Characters"
            />
          </React.Fragment>,
          <ResultList
            list={
              Array [
                "films1",
                "films2",
              ]
            }
            type="Films"
          />,
        ],
        "className": "form-container p-5",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "resistance mb-4",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <label
                className="text-white"
                htmlFor="showPeople"
              >
                Show characters
              </label>,
              <input
                checked={true}
                id="showPeople"
                onClick={[Function]}
                type="checkbox"
              />,
              <br />,
              <label
                className="text-white"
                htmlFor="showFilms"
              >
                Show films
              </label>,
              <input
                checked={true}
                id="showFilms"
                onClick={[Function]}
                type="checkbox"
              />,
              <div
                className="text-white my-3"
              >
                4 RESULTS
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Show characters",
                "className": "text-white",
                "htmlFor": "showPeople",
              },
              "ref": null,
              "rendered": "Show characters",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "checked": true,
                "id": "showPeople",
                "onClick": [Function],
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Show films",
                "className": "text-white",
                "htmlFor": "showFilms",
              },
              "ref": null,
              "rendered": "Show films",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "checked": true,
                "id": "showFilms",
                "onClick": [Function],
                "type": "checkbox",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "4 RESULTS",
                "className": "text-white my-3",
              },
              "ref": null,
              "rendered": "4 RESULTS",
              "type": "div",
            },
          ],
          "type": Symbol(react.fragment),
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Formik
                enableReinitialize={false}
                initialValues={
                  Object {
                    "selectedFilm": "",
                  }
                }
                isInitialValid={false}
                onSubmit={[Function]}
                render={[Function]}
                validateOnBlur={true}
                validateOnChange={true}
              />,
              <Formik
                enableReinitialize={false}
                initialValues={
                  Object {
                    "selectedGender": "",
                  }
                }
                isInitialValid={false}
                onSubmit={[Function]}
                render={[Function]}
                validateOnBlur={true}
                validateOnChange={true}
              />,
              <ResultList
                list={
                  Array [
                    Object {
                      "birth_year": "1234",
                      "eye_color": "eye_color",
                      "gender": "male",
                      "hair_color": "hair_color",
                      "height": "123",
                      "mass": "12",
                      "name": "name",
                    },
                    Object {
                      "birth_year": "1234232",
                      "eye_color": "black",
                      "gender": "female",
                      "hair_color": "black",
                      "height": "12333",
                      "mass": "12333",
                      "name": "name2",
                    },
                  ]
                }
                type="Characters"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "enableReinitialize": false,
                "initialValues": Object {
                  "selectedFilm": "",
                },
                "isInitialValid": false,
                "onSubmit": [Function],
                "render": [Function],
                "validateOnBlur": true,
                "validateOnChange": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "enableReinitialize": false,
                "initialValues": Object {
                  "selectedGender": "",
                },
                "isInitialValid": false,
                "onSubmit": [Function],
                "render": [Function],
                "validateOnBlur": true,
                "validateOnChange": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "list": Array [
                  Object {
                    "birth_year": "1234",
                    "eye_color": "eye_color",
                    "gender": "male",
                    "hair_color": "hair_color",
                    "height": "123",
                    "mass": "12",
                    "name": "name",
                  },
                  Object {
                    "birth_year": "1234232",
                    "eye_color": "black",
                    "gender": "female",
                    "hair_color": "black",
                    "height": "12333",
                    "mass": "12333",
                    "name": "name2",
                  },
                ],
                "type": "Characters",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": Symbol(react.fragment),
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "list": Array [
              "films1",
              "films2",
            ],
            "type": "Films",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
